### ISSUE TITLE
Panic in master process when querying status during upgrade

### ISSUE THREAD

**Issue by roosmaa** [sozu-proxy/sozu#296]
During the upgrade cycle when the workers are created and stopped, the master process can panic when the sozuctl status is executed.

```
thread 'main' panicked at 'invalid index', /checkout/src/libcore/option.rs:839:4                                                                                                                                                                                                          
stack backtrace:                                                                                                                                                                                                                                                                          
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace                                                                                                                                                                                                                            
              at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49                                                                                                                                                                                  
   1: std::sys_common::backtrace::_print                                                                                                                                                                                                                                                  
             at /checkout/src/libstd/sys_common/backtrace.rs:69                                                                                                                                                                                                                           
   2: std::panicking::default_hook::{{closure}}                                                                                                                                                                                                                                           
             at /checkout/src/libstd/sys_common/backtrace.rs:58                                                                                                                                                                                                                           
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
2017-12-19T23:08:44Z 2798966470873955 24825 WRK-04 ERROR        error loading default 404 answer file: None
2017-12-19T23:08:44Z 2798966470901786 24825 WRK-04 ERROR        error loading default 503 answer file: None
2017-12-19T23:08:44Z 2798966483467524 24825 WRK-04 INFO registered listener for app TcpTest on port 8081 at token Token(12297829382473034412)
2017-12-19T23:08:44Z 2798966483584294 24825 WRK-04 ERROR        No front for this instance
2017-12-19T23:08:44Z 2798966483599723 24825 WRK-04 ERROR        Couldn't add tcp instance
2017-12-19T23:08:44Z 2798966483710832 24825 WRK-04 INFO 4 starting event loop
2017-12-19T23:08:44Z 2798966483817276 24825 WRK-04 INFO ID-eEcg0B status
             at /checkout/src/libcore/panicking.rs:71
   9: core::option::expect_failed
             at /checkout/src/libcore/option.rs:839
  10: <core::option::Option<T>>::expect
             at /checkout/src/libcore/option.rs:302
  11: <slab::Slab<T, I> as core::ops::index::IndexMut<I>>::index_mut
             at /home/mart/.cargo/registry/src/github.com-1ecc6299db9ec823/slab-0.3.0/src/lib.rs:283
  12: sozu::command::CommandServer::handle_worker_message
             at src/command/mod.rs:510
  13: sozu::command::CommandServer::handle_worker_events
             at src/command/mod.rs:370
  14: sozu::command::CommandServer::ready
             at src/command/mod.rs:283
  15: sozu::command::CommandServer::run
             at src/command/mod.rs:197
  16: sozu::upgrade::begin_new_master_process
             at src/upgrade.rs:129
  17: sozu::main
             at src/main.rs:97
  18: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  19: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  20: main
  21: __libc_start_main
  22: _start
2017-12-19T23:08:44Z 2798966541338135 24824 WRK-03 ERROR        command channel was closed
2017-12-19T23:08:44Z 2798966541411489 24825 WRK-04 ERROR        error reading from command channel
2017-12-19T23:08:44Z 2798966541533750 24823 WRK-02 ERROR        command channel was closed
```

To reproduce, in one terminal run sozuctl status in a loop:

```
$ while true; do ../target/debug/sozuctl -c config.toml status ; sleep 1; done
```

And then in another terminal run the sozuctl upgrade. With the current master branch upgrade procedure it is slightly more difficult to reproduce, but possible. I'm working on updating the logic in the upgrade command right now and when rolling over the workers one by one, it's fairly easy to reproduce. The upgrade command in commit 1c62d26 should be able to reproduce this bug very reliably.

**Comment by NotBad4U** [comment]
This is related to #275. We have to make a refactoring on the upgrade process.
The master process should not try to answer to the clients during his upgrading process.

**Comment by roosmaa** [comment]
@NotBad4U actually this happens with launch worker and soft stop worker combo. I commented out the actual master process upgrade and it has no effect on this bug.

**Comment by roosmaa** [comment]
Ok, so what I intended to post was the launch/soft-stop crash. Apparently the backtrace in the original post was for a different one (it was late, didn't really double check). The previous backtrace does seem to be related to master upgrade.

Anyway, here's the backtrace of the panic that happens during launch worker + soft stop worker loop, if someone requests for a `sozuctl status` at the same time.

```
2017-12-20T11:14:26Z 2842508284370000 346 WRK-05 INFO   ID-I1pVIT status                                                                                                                                                                                                                  
stack backtrace:                                                                                                                                                                                                                                                                          
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace                                                                                                                                                                                                                            
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49                                                                                                                                                                                                               
   1: std::sys_common::backtrace::_print                                                                                                                                                                                                                                                  
             at /checkout/src/libstd/sys_common/backtrace.rs:69                                                                                                                                                                                                                           
   2: std::panicking::default_hook::{{closure}}                                                                                                                                                                                                                                           
             at /checkout/src/libstd/sys_common/backtrace.rs:58                                                                                                                                                                                                                           
             at /checkout/src/libstd/panicking.rs:381                                                                                                                                                                                                                                     
   3: std::panicking::default_hook                                                                                                                                                                                                                                                        
             at /checkout/src/libstd/panicking.rs:397                                                                                                                                                                                                                                     
   4: std::panicking::rust_panic_with_hook                                                                                                                                                                                                                                                
             at /checkout/src/libstd/panicking.rs:577                                                                                                                                                                                                                                     
   5: std::panicking::begin_panic                                                                                                                                                                                                                                                         
             at /checkout/src/libstd/panicking.rs:538                                                                                                                                                                                                                                     
   6: std::panicking::begin_panic_fmt                                                                                                                                                                                                                                                     
             at /checkout/src/libstd/panicking.rs:522                                                                                                                                                                                                                                     
   7: rust_begin_unwind                                                                                                                                                                                                                                                                   
             at /checkout/src/libstd/panicking.rs:498                                                                                                                                                                                                                                     
   8: core::panicking::panic_fmt                                                                                                                                                                                                                                                          
2017-12-20T11:14:26Z 2842508422065297 347 WRK-06 ERROR  error loading default 404 answer file: None                                                                                                                                                                                       
2017-12-20T11:14:26Z 2842508422101519 347 WRK-06 ERROR  error loading default 503 answer file: None                                                                                                                                                                                       
2017-12-20T11:14:26Z 2842508442294251 347 WRK-06 INFO   registered listener for app TcpTest on port 8081 at token Token(12297829382473034412)
2017-12-20T11:14:26Z 2842508442448913 347 WRK-06 ERROR  No front for this instance
2017-12-20T11:14:26Z 2842508442468709 347 WRK-06 ERROR  Couldn't add tcp instance
2017-12-20T11:14:26Z 2842508442730398 347 WRK-06 INFO   6 starting event loop
2017-12-20T11:14:26Z 2842508442914045 347 WRK-06 INFO   ID-nnhqr0 status
             at /checkout/src/libcore/panicking.rs:71
   9: core::option::expect_failed
             at /checkout/src/libcore/option.rs:839
  10: <core::option::Option<T>>::expect
             at /checkout/src/libcore/option.rs:302
  11: <slab::Slab<T, I> as core::ops::index::IndexMut<I>>::index_mut
             at /home/mart/.cargo/registry/src/github.com-1ecc6299db9ec823/slab-0.3.0/src/lib.rs:283
  12: sozu::command::CommandServer::handle_worker_message
             at src/command/mod.rs:510
  13: sozu::command::CommandServer::handle_worker_events
             at src/command/mod.rs:370
  14: sozu::command::CommandServer::ready
             at src/command/mod.rs:283
  15: sozu::command::CommandServer::run
             at src/command/mod.rs:197
  16: sozu::command::start
             at src/command/mod.rs:603
  17: sozu::main
             at src/main.rs:130
  18: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  19: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  20: main
  21: __libc_start_main
  22: _start
2017-12-20T11:14:26Z 2842508454993532 348 WRK-07 ERROR  error reading from command channel
2017-12-20T11:14:26Z 2842508455134292 347 WRK-06 ERROR  error reading from command channel
2017-12-20T11:14:26Z 2842508455768676 346 WRK-05 ERROR  error reading from command channel
```

**Comment by Geal** [comment]
I cannot reproduce, but this particular panic should be fixed by 37d32f8

**Comment by roosmaa** [comment]
Cherry picked 37d32f8 into my env and indeed it doesn't crash anymore. Though now it seems occasionally `sozuctl status` command gets an unexpected reply from proxy: 

```
received message with invalid id: ConfigMessageAnswer { id: "ID-aHIrny", version: 0, status: Ok, message: "ok: 1 messages, error: {}", data: None }
```

Anyyway, the original panic seems to be addressed, so closing this issue.
